[Ryan Chang Jia Jie] iP

rcjj98 opened the pull request:
# Siri

Siri frees your mind of having to remember things you need to do.

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/rcjj98/ip).
2. ensure you are in the right directory to compile ``` javac siri.Siri``` and run ```java Siri```.
3. add your tasks. 📝 
```
todo run 2.4km
deadline CS2103T quiz /by Friday 2pm
event Jane's Wedding /at Fullerton Hotel
````
4. let it remember your tasks for you. 🧠  

And it's **FREE**!

Features:

- [x] Keep track of tasks 
- [x] Manage deadlines
- [x] Finding specific tasks within the list
- [ ]  Reminders (Coming soon!)

Here's the main method:
```java
public static void main(String[] args) {
        new Siri("data/siri.txt").run();
}
```

Commit:
add: .gitignore

Commit:
add: Siri/Level1.java

Commit:
add: Level1.java

Commit:
add javadocs comments

Commit:
add javadocs comments

Commit:
add Level2.java

Commit:
add Siri/Level2

Commit:
update Level2.java

Commit:
update Siri/Level2

Commit:
add Level3.java

Commit:
neaten 1 line of code in Level2.java

Commit:
add Level3.java

Commit:
Shift Level3.java to L3 folder

Commit:
add Task.java

Commit:
update Level3.java

Commit:
Revert "update Level3.java"

This reverts commit 04d905b72b04ad4c71e6d33caa9c093a7d0e08f3.

Commit:
Revert "Revert "update Level3.java""

This reverts commit 8e13127de05221bdf90dc5ddca600c370691b933.

Commit:
add mark logic to Level3.java

Commit:
add markAsNotDone method to Task.java

Commit:
update Level3.java

Commit:
add javadoc comments to Level3.java

Commit:
update Siri replies to mark and unmark user inputs

Commit:
condense multiple system.out.println into lesser lines in Level3.java

Commit:
Merge commit '38fcffd337170938ce966b3fd4cdb031fb1fb05a'

* commit '38fcffd337170938ce966b3fd4cdb031fb1fb05a':
  update Siri/Level2
  add Siri/Level2
  add javadocs comments
  add: Siri/Level1.java

# Conflicts:
#	Siri/Level2.java

Commit:
move level1.java into L1 folder

Commit:
move Level2.java into L2 folder

Commit:
add Level4.java into L4 folder

Commit:
add Task.java to L4 folder

Commit:
add Todo.java to L4 folder

Commit:
add @Override tag in line 24 of Task.java

Commit:
update Todo.java

Commit:
update Level4.java

Commit:
add Deadline.java to L4 folder

Commit:
update Deadline.java

Commit:
update Level4.java to accept deadline and todo tasks

Commit:
add javadoc comment to line 51

Commit:
update Todo.java toString method

Commit:
update Deadline.java toString method

Commit:
add Event.java to L4 folder

Commit:
update Level4.java string outputs for todo, deadline and event tasks

Commit:
update input.txt

Commit:
update runtest.sh to include path to main class on line 16

Commit:
update Level4.java removed final else condition

Commit:
Move Deadline.java to L5 folder

Commit:
Move Task.java to L5 folder

Commit:
Move Todo.java to L5 folder

Commit:
Move Event.java to L5 folder

Commit:
add EmptyDescException.java

Commit:
Level5.java: update elseif condition for "todo" to handle exception

Commit:
Level5.java: Update if condition to identify whitespaces as empty input

Commit:
Level5.java: Add reply logic to respond to unrecognised tasks/words

Commit:
Level5.java: Update greeting message to include instructions on usage

Commit:
Copy Todo.java over to L6

Commit:
Copy Event.java over to L6

Commit:
Copy Deadline.java over to L6

Commit:
Copy Task.java over to L6

Commit:
Copy EmptyDescException.java over to L6

Commit:
Level6.java: Add the option to delete tasks from list

zunedz commented on Siri/L4/Level4.java:
```diff
@@ -0,0 +1,80 @@
+public class Level4 {
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
+        ArrayList<Task> arrList = new ArrayList<Task>();
```
perhaps changing the variable name to be more meaningful (e.g. taskList, tasks)?

zunedz commented on Siri/L4/Task.java:
```diff
@@ -0,0 +1,29 @@
+        this.isDone = false; //initialise as false
+    }
+
+    public String getStatusIcon() { //will return either "X" or " " depending if task is done or not
```
it's good that you use verb/action to name the methods!

zunedz commented on Siri/L5/Level5.java:
```diff
@@ -0,0 +1,105 @@
+                    checkRestOfInput(desc);
+                    Todo t = new Todo(desc);
+                    arrList.add(t);
+                    System.out.println(line + "     This task has been added to your list:\n       "
```
would it be better if a long sentence/code is wrapped in a new line instead (with proper indentation of course)?

zunedz commented on Siri/L6/Task.java:
```diff
@@ -0,0 +1,29 @@
+public class Task {
+    String description;
+    boolean isDone;
```
variable naming is already on point!

zunedz commented on Siri/L2/Level2.java:
```diff
@@ -0,0 +1,29 @@
+        String line = "    __________________________________________________\n";
+        System.out.println(line + "     Hi there. I'm Siri \n" + "     How may I help you? \n" + line);
+
+        while(true) {
```
Perhaps it would be better to put spaces before the parenthesis? in adherence to the java coding standard.

zunedz commented on Siri/L4/Event.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + "[" + this.getStatusIcon() + "]" + this.description + "(at:" + this.eventDate + ")";
```
Since this line is quite long, any reason why you did not wrap it on a new line?

zunedz commented on Siri/L5/Level5.java:
```diff
@@ -0,0 +1,105 @@
+                + "     What do you wish to take note of today? \n" + line);
+
+        while (true) { //will keep querying user for input until user inputs "bye"
+            String word = sc.next();
```
Perhaps changing the variable name to be more intuitive would be a good idea? For example instruction/command

zunedz commented on Siri/L6/Level6.java:
```diff
@@ -0,0 +1,113 @@
+            } else if (word.equals("mark")) { //marks the specified task with an "X"
+                int num = sc.nextInt();
+                int index = num - 1;
+                Task t = arrList.get(index);
```
Maybe you can consider renaming the variable t to another name, such as taskToMark (or any shorter version)?

zunedz commented on Siri/L6/Level6.java:
```diff
@@ -0,0 +1,113 @@
+            return;
+        }
+    }
+}
```
Just a general suggestion, maybe you can also consider modifying your code structure? I think you can take a look at others' code, the code is structured within the src/java directory, and no need to have a separate folder for different levels as it is handled by git/version control.

zihaowrez commented on Siri/L3/Level3.java:
```diff
@@ -0,0 +1,53 @@
+public class Level3 {
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
+        ArrayList<Task> arrList = new ArrayList<Task>();
```
Should `arrList` have a more meaningful name?

zihaowrez commented on Siri/L1/Level1.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Scanner;
+
+public class Level1 {
```
Perhaps adding a Javadoc for each class and public method?

tanweien replied:
I agree that adding a Javadoc for each class and public method would allow the programme to be better understood

zihaowrez commented on Siri/L4/Level4.java:
```diff
@@ -0,0 +1,80 @@
+                            + line);
+                    break;
+                }
+            } else { } //do nothing
```
Maybe the last `else { }` can be omitted?

zihaowrez commented on Siri/L5/Level5.java:
```diff
@@ -0,0 +1,105 @@
+                //creates event task with "[E]" prefix and event date postfix and adds to list
+                while (sc.hasNextLine()) {
+                    String[] userInput = sc.nextLine().split("/at");
+                    // The userInput array now contains [taskDescription, event date]
```
I like this comment, which is helpful for reviewers and compliant with the coding standard

zihaowrez reviewed (commented):
Most of the code is in good shape with respect to the coding standard, and adding the Javadocs would make it more complete.

I also second @zunedz 's suggestion, that there's no need to create a folder for each level. You can work on the same files and tag your commit for each level in git.

tanweien commented on Siri/L4/Level4.java:
```diff
@@ -0,0 +1,80 @@
+                while (sc.hasNextLine()) {
+                    String[] userInput = sc.nextLine().split("/by");
+                    // The userInput array now contains [taskDescription, deadline]
+                    Deadline deadLineTask = new Deadline(userInput[0], userInput[1]);
```
Good job on the clean code, however it might be better to avoid using magic numbers for better readability!

tanweien commented on Siri/L1/Level1.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Scanner;
+
+public class Level1 {
```
It might be better to use a more intuitive class name here! I noticed the same issue in other places too.

tanweien commented on Siri/L3/Level3.java:
```diff
@@ -0,0 +1,53 @@
+        System.out.println(line + "     Hi there. I'm Siri \n" + "     How may I help you? \n" + line);
+
+        while (true) { //will keep querying user for input until user inputs "bye" into system
+            String word = sc.next();
```
Perhaps a more meaningful variable name would be easier to understand

tanweien commented on Siri/L3/Task.java:
```diff
@@ -0,0 +1,28 @@
+public class Task {
+    String description;
+    boolean isDone;
```
Nice naming boolean convention that follows recommended guidelines

tanweien commented on Siri/L5/Task.java:
```diff
@@ -0,0 +1,29 @@
+public class Task {
```
Good job adding comments to each method giving a brief explanation. However, perhaps adding Javadoc might be better!

Commit:
update Task.java to include first letter of task as an attribute of task

Commit:
update Todo.java constructor method

Commit:
update Deadline.java constructor and toString() method

Commit:
update Event.java constructor and toString() methods

Commit:
move EmptyDescException.java from L6 to L7 folder

Commit:
Add Siri.txt

Commit:
Update Level7.java to write/read to/from Siri.txt

Commit:
Neaten a few lines of code in Level7.java

Commit:
Add comment in Level7.java at line 35

Commit:
Update few lines of code in Level6.java

Commit:
Copy Task.java from L6 to L8 folder

Commit:
Copy EmptyDescException.java over from L6 to L8 folder

Commit:
Copy Todo.java over from L6 to L8 folder

Commit:
Overload Deadline.java constructor method to take in dates and dateTime

Commit:
Overload Event.java constructor method inputs to accept date or dateTime

Commit:
Update Level8.java to understand dates and times in deadline/event tasks

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update Level8.java to understand dates and times in deadline/event tasks
  Overload Event.java constructor method inputs to accept date or dateTime
  Overload Deadline.java constructor method to take in dates and dateTime
  Copy Todo.java over from L6 to L8 folder
  Copy EmptyDescException.java over from L6 to L8 folder
  Copy Task.java from L6 to L8 folder

Commit:
Add Storage.java to load/save from/to the specified text file

Commit:
Add TaskList.java to handle adding/deleting tasks in the list

Commit:
Update EmptyDescException.java name to InvalidInputException.java

Commit:
Add Ui.java to handle interactions with the user

There are several different required outputs to print to ensure
interactions with the user make sense to them.

Consolidating the majority of the print statements into Ui.java instead
of lumping them with the main method improves code quality and follows
coding standards.

As a step towards improving code quality and utilising OOP, let's
abstract the code that handles print statements into Ui.java. Doing so
will make code readability and quality better.

Commit:
Add Parser.java to handle the logic of the program when interacting with
the user

There are several different valid commands the program can take as
inputs, in some cases the commands are not inputted in a valid syntax.

Consolidating the majority of the program logic into Parser.java instead
of lumping them with the main method improves code quality and follows
coding standards.

As a step towards improving code quality and utilising OOP, let's
abstract the code that handles the program logic into Parser.java. Doing
so will make code readability and quality better.

Commit:
Add Command.java as super class to extract more classes for the different
types of commands

Commit:
Add AddDeadlineCommand.java to create new deadline tasks to be added into
list

Commit:
Update AddDeadlineCommand.java to add new line of code at the end of the
file

Commit:
Add AddEventCommand.java & AddTodoCommand.java to create new tasks to be
added into list

Commit:
Add MarkCommand.java & UnmarkCommand.java to mark/unmark tasks

Commit:
Add DeleteCommand.java & ListCommand.java to delete & list tasks

Commit:
Add ExitCommand.java to exit the program

Commit:
Move Deadline.java, Event.java, Todo.java & Task.java over to
src/main/java

Commit:
Add Siri.java to call all necessary methods to get Siri program to run

Commit:
Add siri.txt file to read/write task details in

Commit:
Delete Siri Folder containing level1 to level8 of iP

Commit:
Package all classes into siri in src/main/java/siri

Commit:
Remove some lines of unreachable code

Commit:
Add DeadlineTest.java, TodoTest.java & EventTest.java to test behavior of
code

Commit:
ignore MANIFEST.MF file

Commit:
Add JavaDocs comments to the code

Commit:
Tweak the code to comply with coding standards

Commit:
Add FindCommand.java to handle finding of tasks with specified keyword

Commit:
Add Gradle support to iP

Commit:
Add a GUI to Siri

Commit:
Add Assertions to TaskList.java

There are some user inputs that could result in unexpected/undesirable
behaviours of the program.

Asserting those inputs ensures the correct program behaviour will be
reproduce.

As a step towards maintaining the right program response, let's use
the assert feature to document crucial assumptions that needs to hold at
various points in the code

Commit:
Improve code quality

There are many areas of code quality to ensure that one's program is
considered a fine piece of work.

Amongst the various dimensions of code quality, such as run-time
efficiency, security, and robustness, one of the most important is
understandability.

Thus let us make an effort to maintain code quality not just for others
to understand, but also for ourselves as we might become 'strangers'
to our own code someday.

Commit:
Add C-Help extension to siri

Commit:
Improve code quality

rcjj98 committed:
Merge pull request #2 from rcjj98/branch-A-Assertions

Add Assertions to TaskList.java

Commit:
Merge branch 'master' into branch-C-Help

* master:
  Improve code quality

# Conflicts:
#	src/main/java/siri/Parser.java

rcjj98 committed:
Merge pull request #4 from rcjj98/branch-C-Help

Add C-Help extension to siri

Commit:
Update both user and program profile pictures

Commit:
Add Ui.png to docs directory

Commit:
Give credit to peer for code reusage

rcjj98 committed:
Set theme jekyll-theme-tactile

Commit:
Add a User Guide

Commit:
Merge branch 'master' of https://github.com/rcjj98/ip

* 'master' of https://github.com/rcjj98/ip:
  Set theme jekyll-theme-tactile

rcjj98 committed:
Set theme jekyll-theme-cayman

Commit:
Update User Guide to be neater

Commit:
Merge branch 'master' of https://github.com/rcjj98/ip

* 'master' of https://github.com/rcjj98/ip:
  Set theme jekyll-theme-cayman

Commit:
Update Storage.java to write/load to/from the correct directory

Commit:
Update UserGuide

Commit:
Update UserGuide

Commit:
Update UserGuide in README.md file

Commit:
Remove empty comment in Storage.java

Commit:
Add header comments

Commit:
Add header comments

Commit:
Add header comments

Commit:
Update header comments to follow Javadoc comments coding standards
